transport/src/main/java/com/strategyobject/substrateclient/transport/ws/BackoffReconnectionPolicy.java

```diff
@@ -0,0 +1,126 @@
+package com.strategyobject.substrateclient.transport.ws;
+
+import com.google.common.base.Preconditions;
+import lombok.*;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Represents a backoff retry policy
+ */
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@Getter
+@Slf4j
+public class BackoffReconnectionPolicy implements ReconnectionPolicy {
+    /**
+     * The number of executed attempts to reconnect
+     */
+    private int counter = 0;
+    /**
+     * Max number of attempts
+     */
+    private final int maxAttempts;
+    /**
+     * Initial delay
+     */
+    private final long delay;
+    /**
+     * Time duration at a given unit
+     */
+    private final TimeUnit unit;
+    /**
+     * Max delay
+     */
+    private final long maxDelay;
+    /**
+     * A multiplier that's applied to delay after every attempt
+     */
+    private final double factor;
+
+    /**
+     * @param context contains a reason of disconnection
+     * @return a schedule for the next reconnection
+     */
+    @Override
+    public @NonNull ReconnectionSchedule apply(@NonNull ReconnectionContext context) {
+        try {
+            if (counter >= maxAttempts) {
+                log.info("Provider won't reconnect more.");
+
+                return ReconnectionSchedule.NEVER;
+            }
+
+            var nextDelay = delay * Math.pow(factor, counter);
```

Comment:
As far as I can see negatives and numbers less than one are allowed, which might result in a tricky behavior. Is it intentional?

Reply:
Negatives are not allowed there. I left an opportunity to use numbers less than one intentionally, but probably it doesn't make much sense.